source/libip4string/stoip4.c

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright (c) 2014-2018 ARM Limited. All rights reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ * Licensed under the Apache License, Version 2.0 (the License); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an AS IS BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include <string.h>
+#include <stdlib.h>
+#include <stdint.h>
+#include "common_functions.h"
+#include "ip4string.h"
+
+/**
+ * Convert numeric IPv4 address string to a binary.
+ * \param ip4addr IPv4 address in string format.
+ * \param len Length of IPv4 string, maximum of 16..
+ * \param dest buffer for address. MUST be 4 bytes.
+ * \return boolean set to true if correction succeded, false if it didn't
+ */
+bool stoip4(const char *ip4addr, size_t len, void *dest)
+{
+    uint8_t *addr = dest;
+
+    if (len > 16) { // Too long, not possible
+        return false;
+    }
+
+    uint_fast8_t stringLength = 0;
+
+    //Iterate over each component of the IP
+    for (uint_fast8_t byteCount = 0; byteCount < 4; byteCount++) {
+
+        //No valid character (IPv4 addresses don't have implicit 0, that is x.y..z being read as x.y.0.z)
+        if(stringLength == len || ip4addr[stringLength] < '0' || ip4addr[stringLength] > '9') {
+            return false;
+        }
+
+        //For each component, we convert it to the raw value
+        uint_fast16_t byte = 0;
+        while(stringLength < len && ip4addr[stringLength] >= '0' && ip4addr[stringLength] <= '9') {
+            byte *= 10;
+            byte += ip4addr[stringLength++] - '0';
+
+            //We go over the maximum value for an IPv4 component
+            if(byte > 0xff) {
+                return false;
+            }
+        }
+
+        //Append the component
+        addr[byteCount] = (uint8_t) byte;
+
+        //If we're at the end or if the next character is invalid, we stop
+        // This is expected if we just dealt with the last component in which case we can return true
+        if(stringLength == len || !ip4addr[stringLength] || ip4addr[stringLength] != '.') {
```

Comment:
Hmm, still not quite right. Hitting `len` or NUL is okay if byteCount == 3, but not something else. And a `.` would be bad if byteCount was == 3.

Hows about 

    if (byteCount == 3) {
        break;
    }
    if (stringLength == len || ip4[stringLength++] != '.') {
        return false;
    }

And after the loop

    return stringLength == len || ip4[stringLength] == '\0';

`stoip6` is a bit woolly on the specified length thing, but I think this makes it roughly consistent. It was intended to allow the caller to have its own special delimiter prechecked (eg '/' for prefix, or ']' for URL).

Reply:
I'm really not a big fan of `if (byteCount == 3)` as it makes it harder to determine where the loop will end (and we may as well turn the `for` loop in an infinite one, as its breaking condition will never trigger).

However, "cleaner" structures I can think of look like the following which again obfuscate where the legitimate exit point is.
```
    if (stringLength == len) {
        return byteCount == 3;
    }

    if (ip4addr[stringLength++] != '.') {
        return false;
    }
}
return false;
```
Looks like we have the choice of either turning the `for` loop in an infinite loop and keep the return after the loop, or move it in but make the "default" return a `return false`.

I never thought there could be so much head scratching on an IPv4 parser!

Reply:
Yes, it may as well be an infinite loop. May be the way to do it.  Not too bothered exactly how it is written, as long as we can rely on the "true/false" result to be the accurate "is this an IPv4 literal" check that we need.